internal.nix

```diff
@@ -62,27 +62,74 @@ rec {
   # Type: Path -> Derivation
   patchedLockfile = file: writeText "packages-lock.json" (builtins.toJSON (patchLockfile file));
 
+  # Prepared source for npm installation
+  nodeSource = nodes: runCommand "node-sources-${nodejs.version}"
+    { } ''
+    tar --no-same-owner --no-same-permissions -xf ${nodejs.src}
+    mv node-* $out
+  '';
+
   node_modules =
     { src
     , packageJson ? src + "/package.json"
     , packageLockJson ? src + "/package-lock.json"
     , buildInputs ? [ ]
     , nativeBuildInputs ? [ ]
     , nodejs ? default_nodejs
+    , preBuild ? ""
+    , postBuild ? ""
+    , preInstallLinks ? { } # set that describes which files should be linked in a specific packages folder
     , ...
     }@args:
     let
       lockfile = readLockfile packageLockJson;
+
+      preinstall_node_modules = writeTextFile {
+        name = "prepare";
+        destination = "/node_modules/.hooks/prepare";
+        text = ''
+          #! ${stdenv.shell}
+
+          ${lib.concatStringsSep "\n" (
+            lib.mapAttrsToList (name: mappings: ''
+                if [ "$npm_package_name" == "${name}" ]; then
+                ${lib.concatStringsSep "\n"
+                  (lib.mapAttrsToList (to: from: ''
+                        dirname=$(dirname ${to})
+                        mkdir -p $dirname
+                        ln -s ${from} ${to}
+                      '') mappings
+                  )}
+                fi
+              '') preInstallLinks
+            )}
+
+          if grep -I -q -r '/bin/' .; then
```

Comment:
This is a bit of a dirty hack but it gave me a >6x speedup when installing all the dependencies. Runtime went from ~7 minutes to around 1 minute.